subsys/bluetooth/host/settings.c

```diff
@@ -119,6 +119,14 @@ static int set(const char *name, size_t len_rd, settings_read_cb read_cb,
 	ssize_t len;
 	const char *next;
 
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_ENABLE)) {
```

Comment:
Does the Bluetooth's settings `set` handler actually require that Bluetooth is enabled before? I think not, but maybe I missed something here (please correct me if I am wrong).
If it's not required then we may handle settings set and then just delay calling commit to ensure that it's called after bt_enable. That approach would be simpler - no matter when user calls settings_load, the Bluetooth settings will eventually be handled properly.

Reply:
> Does the Bluetooth's settings set handler actually require that Bluetooth is enabled before? I think not, but maybe I missed something here (please correct me if I am wrong).

Not currently. But I was worried a future change may assume that `bt_enable` has been called. And we may as well skip handling the `set` events if we are not going to handle the `commit`.

> If it's not required then we may handle settings set and then just delay calling commit to ensure that it's called after bt_enable. That approach would be simpler - no matter when user calls settings_load, the Bluetooth settings will eventually be handled properly.

Do you mean that `commit` should set a flag for `bt_enable` to continue the "commit"? Sounds reasonable.

Reply:
> > Does the Bluetooth's settings set handler actually require that Bluetooth is enabled before? I think not, but maybe I missed something here (please correct me if I am wrong).
> 
> Not currently. But I was worried a future change may assume that `bt_enable` has been called. And we may as well skip handling the `set` events if we are not going to handle the `commit`.

IMO it is a good idea to check for the `bt_enable` flag, some things might not be set-up to process the settings entries that are stored. If anything calls the `set` method when `bt` is not enabled it should return without causing problems.

The same holds for the commit.

Reply:
@MarekPieta 

@alwa-nordic writes in his commit message that loading settings send HCI commands to the controller. When calling `bt_enable()` we send the `HCI Reset` opcode to controller, so even if it didn't crash, `bt_enable` would erase any data previously sent to the controller.

Reply:
> > Does the Bluetooth's settings set handler actually require that Bluetooth is enabled before? I think not, but maybe I missed something here (please correct me if I am wrong).
> 
> Not currently. But I was worried a future change may assume that `bt_enable` has been called. And we may as well skip handling the `set` events if we are not going to handle the `commit`.
> 
> > If it's not required then we may handle settings set and then just delay calling commit to ensure that it's called after bt_enable. That approach would be simpler - no matter when user calls settings_load, the Bluetooth settings will eventually be handled properly.
> 
> Do you mean that `commit` should set a flag for `bt_enable` to continue the "commit"? Sounds reasonable.

I think it may be a good approach (it may be simpler from user's perspective and less error prone if user does not need to explicitly handle the special case where settings are loaded before Bluetooth is enabled).

Reply:
> @MarekPieta
> 
> @alwa-nordic writes in his commit message that loading settings send HCI commands to the controller. When calling `bt_enable()` we send the `HCI Reset` opcode to controller, so even if it didn't crash, `bt_enable` would erase any data previously sent to the controller.

Settings commit callbacks sends the HCI commands. Settings set callback does not do it currently, but as mentioned by @alwa-nordic it could do it at some point in the future.

Reply:
> Settings commit callback sends the HCI commands. Settings set callback does not do it currently, but as mentioned by @alwa-nordic it could do it at some point in the future. If it's the case then I understand that solution proposed by me may be rejected. Still I think that, if it's possible, we should ensure that Bluetooth settings would be automatically loaded at some point after settings_load is called globally to make solution less error prone.

Would it make sense to conditionally automatically load settings in `bt_enable`? Either via a Kconfig option, or a new argument to the function?

Reply:
@Thalley, from CONFIG_BT_SETTINGS ducumentation:
```
When this option has been enabled, it's important that the
application makes a call to settings_load() after having done
all necessary initialization (e.g. calling bt_enable). The
reason settings_load() is handled externally to the stack, is
that there may be other subsystems using the settings API, in
which case it's more efficient to load all settings in one go,
instead of each subsystem doing it independently.
```
In general, Zephyr used to assume that we load all of the settings globally, assuming that modules are prepared for this (this was done for efficiency). Automatically loading settings in bt_enable would contradict a bit with this assumption (it would be a "special case"). That's why I was thinking about remembering values during set callback and then only ensuring that commit is called after the bt_enable to prevent errors.

Reply:
I would like to keep the current approach in this PR as the bug fix. It is compatible with the API doc. We can open an enhancement discussion on improving the API.